pageserver/src/tenant/vectored_blob_io.rs

```diff
@@ -76,14 +76,115 @@ pub(crate) enum VectoredReadExtended {
     No,
 }
 
-pub(crate) struct VectoredReadBuilder {
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum VectoredReadCoalesceMode {
+    /// Only coalesce exactly adjacent reads.
+    AdjacentOnly,
+    /// In addition to adjacent reads, also consider reads whose corresponding
+    /// `end` and `start` offsets reside at the same chunk.
+    Chunked(usize),
+}
+
+impl VectoredReadCoalesceMode {
+    pub(crate) fn get() -> Self {
+        let align = virtual_file::get_io_buffer_alignment();
+        if align == 1 {
+            VectoredReadCoalesceMode::AdjacentOnly
+        } else {
+            VectoredReadCoalesceMode::Chunked(align)
+        }
+    }
+}
+
+pub(crate) enum VectoredReadBuilder {
+    Adajacent(AdjacentVectoredReadBuilderInner),
+    Chunked(ChunkedVectoredReadBuilderInner),
+}
+
+impl VectoredReadBuilder {
+    pub(crate) fn new(
+        start_offset: u64,
+        end_offset: u64,
+        meta: BlobMeta,
+        max_read_size: usize,
+        mode: VectoredReadCoalesceMode,
+    ) -> Self {
+        match mode {
+            VectoredReadCoalesceMode::AdjacentOnly => {
+                Self::Adajacent(AdjacentVectoredReadBuilderInner::new(
+                    start_offset,
+                    end_offset,
+                    meta,
+                    Some(max_read_size),
+                ))
+            }
+            VectoredReadCoalesceMode::Chunked(chunk_size) => {
+                Self::Chunked(ChunkedVectoredReadBuilderInner::new(
+                    start_offset,
+                    end_offset,
+                    meta,
+                    Some(max_read_size),
+                    chunk_size,
+                ))
+            }
+        }
+    }
+
+    pub(crate) fn new_streaming(
+        start_offset: u64,
+        end_offset: u64,
+        meta: BlobMeta,
+        mode: VectoredReadCoalesceMode,
+    ) -> Self {
+        match mode {
+            VectoredReadCoalesceMode::AdjacentOnly => Self::Adajacent(
+                AdjacentVectoredReadBuilderInner::new(start_offset, end_offset, meta, None),
+            ),
+            VectoredReadCoalesceMode::Chunked(chunk_size) => {
+                Self::Chunked(ChunkedVectoredReadBuilderInner::new(
+                    start_offset,
+                    end_offset,
+                    meta,
+                    None,
+                    chunk_size,
+                ))
+            }
+        }
```

Comment:
Bit inelegant & bloaty; why not one `new_impl` to which both `new` and `new_streaming` forward?

The only difference is the `Some()` vs `None` in the bodies.
That's the important difference you want the reader to be able to spot easily.

Reply:
Thanks for the advice! 124c97fe6847dc06ecd41e1a387bb0071aa93c96